### PR TITLE
SITES-498 Transform paths in CSS Injector files

### PR DESCRIPTION
**Pull request by jbickar**
# READY FOR REVIEW!!!!

# Summary
- Ostensibly for transforming paths in CSS Injector files. In actuality, it's a full audit and a bit of re-work of the existing find-and-replace logic.

# Needed By (Date)
- Well it's not going to be by the end of the sprint that ends today. So I guess before we migrate any more sites.

# Criticality
- How critical is this PR on a 1-10 scale? 8
- It's pretty big

# Steps to Test

1. Check out this branch
2. Use `sws-migration-testing vhost="sws-oddball-vhost"` in an inventory
3. Optionally add other GSE sites that we had issues with. TELOS, gsediversity, edarchive, crceducation?
4. Run the `migration-playbook.yml` to the `test` environment
5. Go to https://swsmigrationtesting-test.cardinalsites.stanford.edu/ (your site for comparison is https://sws-oddball-vhost.stanford.edu/)
6. Verify links and images on homepage have the correct paths set
7. Go to https://swsmigrationtesting-test.cardinalsites.stanford.edu/css
8. Verify that all 8 divs have a background image
9. Go to https://swsmigrationtesting-test.cardinalsites.stanford.edu/admin/config/development/css-injector/edit/2
10. Verify that all paths in CSS Injector rules have been replaced correctly
11. Go to https://swsmigrationtesting-test.cardinalsites.stanford.edu/view
12. Verify that all links are correct.
    1. *Exception*: Relative links to `href="/sws-migration-test/"` will not be replaced, because we are [skipping replacement of those if the vhost is defined](https://github.com/SU-SWS/ansible-playbooks/blob/SITES-498-css-injector/roles/change-paths/tasks/main.yml#L125).
    2. Ratify the logic of that decision
13. Go to https://swsmigrationtesting-test.cardinalsites.stanford.edu/admin/structure/views/view/view/edit/ and verify link replacement in the header and footer
14. Go to https://swsmigrationtesting-test.cardinalsites.stanford.edu/js
15. Grant yourself 4 staplers
16. Go to https://swsmigrationtesting-test.cardinalsites.stanford.edu/admin/config/development/js-injector/list/js_migration_test/edit
17. Verify link replacement (for single and double-quoted strings!!)
18. Go to https://swsmigrationtesting-test.cardinalsites.stanford.edu/events/sarl
19. Verify that links all are correct
20. Go to https://swsmigrationtesting-test.cardinalsites.stanford.edu/admin/structure/menu/manage/main-menu
21. Verify all of the links in the "sarm" Main Menu item
22. Take a walk through the changes to the code and the (extensive comments). This part of the migration depends heavily on order-of-operations, and I have moved some things around a bit.
23. Review my comments on this PR
24. Review the tags in the `migration-playbook.yml`
25. Celebrate 🎉 

# Affected Projects or Products
- Does this PR impact any particular projects, products, or modules? Sites 2.0

# Associated Issues and/or People
- JIRA ticket: https://stanfordits.atlassian.net/browse/SITES-498 (and a few others, I'm sure)
- Other PRs:
    - #29 
    - #33 
- Anyone who should be notified? ( @sherakama )

# See Also
- [PR Checklist](https://gist.github.com/sherakama/0ba17601381e3adbe0cad566ad4d80a5)

### CODE REVIEW

File: roles/change-paths/defaults/main.yml [hunk 1]
```diff
@@ -1,4 +1,12 @@
 ---
-sar_commands: [ 'sar', 'sarl', 'sarm' ]
+all_sar_commands: [ 'sar', 'sarl', 'sarm', 'sarv', 'sarjs' ]
+sarl_commands: [ 'sarl', 'sarm' ]
+sar_fulltext_commands: [ 'sar', 'sarv', 'sarjs' ]
```
Comment (jbickar) [inline roles/change-paths/defaults/main.yml]: I thought these got merged in #28. Hm.

File: roles/change-paths/tasks/main.yml [hunk 2]
```diff
@@ -46,114 +42,78 @@
...
-
-- name: Debug absolute path for site
-  debug:
-    msg: "Absolute path: {{ new_absolute_path }}"
-
-# Ex: drush @acsf.dev.cardinald7.g2scd7
-- name: Set drush alias
-  set_fact:
-    drush_alias: "drush @acsf{{ drush_environment }}.{{ stack }}.{{ acsf_site_name }}"
-
```
Comment (jbickar) [inline roles/change-paths/tasks/main.yml]: I moved all of this stuff to the new `define-paths` role.

File: roles/change-paths/tasks/main.yml [hunk 3]
```diff
@@ -46,114 +42,78 @@
...
 # See discussion in https://github.com/SU-SWS/ansible-playbooks/pull/17
 # And https://github.com/SU-SWS/ansible-playbooks/pull/22
 # TODO: Determine how to treat for final migration (SITES-465)
-# Ex: drush sar -y 'https://sites.stanford.edu/g2scd7' 'https://g2sc-dev.cardinalsites.stanford.edu/'
+# Ex: drush sar -y 'https://sites.stanford.edu/g2scd7/' 'https://g2sc-dev.cardinalsites.stanford.edu/'
 - name: Update full paths in text
-  shell: "{{ drush_alias }} sar -y '{{ item }}{{ absolute_path_sites_service }}/' 'https://{{ new_absolute_path }}/'"
-  with_items:
-   - "{{ protocols }}"
+  shell: "{{ drush_alias }} {{ item[0] }} -y '{{ item[1] }}{{ absolute_path_sites_service }}/' 'https://{{ new_absolute_path }}/'"
```
Comment (jbickar) [inline roles/change-paths/tasks/main.yml]: I think we must have missed running `{{ sar_fulltext_commands }}` in #28

File: roles/change-paths/tasks/main.yml [hunk 4]
```diff
@@ -167,36 +127,29 @@
...
-
-- name: Print file_public_path
-  debug:
-    msg: "File public path: {{ file_public_path.stdout }}"
-
-- name: Be sure response does not include drush warnings
-  fail:
-    msg: "There appear to be drush warning messages cluttering the file_public_path variable"
-  when: file_public_path.stdout | search('The following module is missing')
-
```
Comment (jbickar) [inline roles/change-paths/tasks/main.yml]: This stuff gut moved to `define-paths` also

File: roles/change-paths/tasks/main.yml [hunk 5]
```diff
@@ -167,36 +127,29 @@
...
-# the ACSF file_public_path value.
+# the ACSF file_public_path value. We probably don't need to loop through
+# {{ protocols }} here (http and https), as we should have already handled that
+# in "Update full paths in text" or "Update full vhost paths in text". But it's
+# cheap to loop through.
 # TODO: Determine how to treat for final migration (SITES-465)
+# Ex. drush sar -y 'https://g2sc.cardinalsites.stanford.edu/sites/default/files' 'https://g2sc.cardinalsites.stanford.edu/sites/g/sbiybj5000986/f'
 - name: Replace absolute sites/default/files with path to files directory
-  shell: "{{ drush_alias }} sar -y '{{ item[0] }}{{ item[1] }}{{ new_absolute_path }}/sites/default/files' '{{ item[0] }}{{ item[1] }}{{ new_absolute_path }}/{{ file_public_path.stdout | replace(' ','') }}'"
+  shell: "{{ drush_alias }} {{ item[0] }} -y '{{ item[1] }}{{ new_absolute_path }}/sites/default/files' 'https://{{ new_absolute_path }}/{{ file_public_path.stdout | replace(' ','') }}'"
```
Comment (jbickar) [inline roles/change-paths/tasks/main.yml]: This fixes a minor bug.

File: roles/css-injector/defaults/main.yml [hunk 6]
```diff
@@ -0,0 +1,8 @@
+---
+all_sar_commands: [ 'sar', 'sarl', 'sarm', 'sarv', 'sarjs' ]
+sarl_commands: [ 'sarl', 'sarm' ]
+sar_fulltext_commands: [ 'sar', 'sarv', 'sarjs' ]
+protocols: [ 'https://', 'http://' ]
+link_attributes:
+  - url\("
+  - url\('
```
Comment (jbickar) [inline roles/css-injector/defaults/main.yml]: Ansible seems to handle the escaping of the quotation marks here. I can't think of another attribute besides `url("https://foo.stanford.edu/sites/default/files...`

File: roles/css-injector/tasks/main.yml [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+# ALTERNATIVE ROLES:
+# --
+#
+# REQUIREMENTS:
+# --
+#
+# KNOWN ISSUES:
+# --
+
+##############################################################################
```
Comment (jbickar) [inline roles/css-injector/tasks/main.yml]: This was all painstakingly re-purposed from `change-paths`. Review for errors. Order-of-operations is important.

File: roles/setup-server/tasks/main.yml [hunk 8]
```diff
@@ -63,3 +63,13 @@
   git:
     repo: "https://github.com/jrglasgow/drush_active_users"
     dest: "~/.drush/drush_active_users"
+
+- name: Update .gitconfig
```
Comment (jbickar) [inline roles/setup-server/tasks/main.yml]: yeah, this was definitely in #28. WTF, Github?

File: roles/change-paths/tasks/main.yml [hunk 9]
```diff
@@ -225,5 +166,8 @@
     - "{{ sarl_commands }}"
   notify: Clear site cache
 
+# TODO: This should not point to livedev and it should point to the correct
```
Comment (jbickar) [inline roles/change-paths/tasks/main.yml]: Must fix this before merging.

File: migration-playbook.yml [hunk 10]
```diff
@@ -35,11 +35,11 @@
   roles:
     - { role: setup-local }
     - { role: download-site, tags: [download-site] }
-    - { role: check-php, tags: [download-site,check-php] }
+    - { role: css-injector, tags: [download-site,css-injector] }
+    - { role: check-php, tags: [download-site,css-injector,check-php] }
```
Comment (jbickar) [inline migration-playbook.yml]: Should `css-injector` be a tag here? What other roles should get the `css-injector` tag?

File: roles/change-paths/defaults/main.yml [hunk 11]
```diff
@@ -3,4 +3,10 @@ all_sar_commands: [ 'sar', 'sarl', 'sarm', 'sarv', 'sarjs' ]
 sarl_commands: [ 'sarl', 'sarm' ]
 sar_fulltext_commands: [ 'sar', 'sarv', 'sarjs' ]
 protocols: [ 'https://', 'http://' ]
-link_attributes: [ 'url="', 'href="', 'src="' ]
+link_attributes:
+  - url="
```
Comment (jbickar) [inline roles/change-paths/defaults/main.yml]: Actually, I'm not sure of the use case of `url=` here. We have:
- `<img src="...`
- `<a href="...`

But what is the `url` attribute? Did we add this when we thought that `sar` was doing find-and-replace on CSS files?

There's also this: https://stackoverflow.com/questions/2725156/complete-list-of-html-tag-attributes-which-have-a-url-value, but I think before we get to that point that we say, "That's it. This is how fine-grained our sifter is; anything else is an edge case."

File: roles/change-paths/tasks/main.yml [hunk 12]
```diff
@@ -214,6 +155,19 @@
     - "{{ link_attributes }}"
   notify: Clear site cache
 
+# We already replaced relative references in href and src attributes (like
+# src="/methane/), but we only did it if the vhost is defined, because of the
+# edge case where inventory_hostname also is a legitimate path for an hreft.
+# Here we more broadly replace the "/inventory_hostname/sites/default/files" string
+# with the ACSF file_public_path value.
+# Ex. drush sar -y 'src="/methane/sites/default/files' 'src="/sites/g/sbiybj5000986/f'
+- name: Replace relative /inventory_hostname/default/files paths with path to files directory
```
Comment (jbickar) [inline roles/change-paths/tasks/main.yml]: Validate the logic here, compared with the previous "Remove relative paths for href, src, and url attributes" task, that only runs when `vhost` is defined

**Comment by jbickar** [comment]
OK, I think the only outstanding issues are:

1. Single quotes in the `{{ link_attributes }}` variables (in both the `change-paths` and `css-injector` roles)
2. Escaping the opening parenthesis in the `{{ link_attributes }}` variable in the `css-injector` role.

File: migration-playbook.yml (hunk 10 above)
Comment (jbickar) [reply migration-playbook.yml]: These tags all need to be reviewed.

Comment (kbrownell) [reply migration-playbook.yml]: Agreed!  Have you been using them?  Perhaps we can do without them entirely?  If you are using them, maybe trim them down?  I've been pretty much just restarting migrations from scratch when something goes wrong.

Comment (jbickar) [reply migration-playbook.yml]: I’ve been running everything from scratch, too. Sounds like this needs to be its own ticket. That’s a good enough answer for me.

I moved things around enough in this PR that I don’t think it should be held up on tags.

Comment (kbrownell) [reply migration-playbook.yml]: Here you go: https://stanfordits.atlassian.net/browse/SITES-594.  Phew.  I'm seeing just how much easier it is to create tickets than to complete them.  That seems like a dangerous dichotomy.

**Comment by jbickar** [comment]
@kbrownell I edited some of my comments on this PR to indicate their current (lack of) relevance. Refresh to get the latest.

File: roles/change-paths/tasks/main.yml [hunk 13]
```diff
@@ -109,12 +65,11 @@
     - "{{ protocols }}"
   notify: Clear site cache
 
+# Note: I don't think this will ever get run, because we've already replaced
```
Comment (jbickar) [inline roles/change-paths/tasks/main.yml]: I created https://stanfordits.atlassian.net/browse/SITES-596 for this and the other one below.

**Comment by kbrownell** [comment]
6. Verify links and images on homepage have the correct paths set - PASS
8. Verify that all 8 divs have a background image - PASS
10. Verify that all paths in CSS Injector rules have been replaced correctly - BEAUTIFUL
12. Verify that all links are correct - PASS
12.ii. I'm still ok with that assumption.
15. Grant yourself 4 staplers - PASS
17. Verify link replacement (for single and double-quoted strings!!) - LOVELY
19. Verify that links all are correct - PERFECT
21. Verify all of the links in the "sarm" Main Menu item - CORRECT

When logged in, drush sends to be the site at https://swsmigrationtesting-test.cardinalsites.stanford.edu and so all relative links are pointing to this non-vhost address.  Once I logged out and access the site again through https://odd-ball-vhost.stanford.edu, then the relative links return to this vhost.  Just a note, I was confused by it for a second.

File: roles/change-paths/tasks/main.yml [hunk 14]
```diff
@@ -214,6 +163,19 @@
     - "{{ link_attributes }}"
   notify: Clear site cache
 
+# We already replaced relative references in href and src attributes (like
+# src="/methane/), but we only did it if the vhost is defined, because of the
+# edge case where inventory_hostname also is a legitimate path for an href.
+# Here we more broadly replace the "/inventory_hostname/sites/default/files" string
```
Comment (kbrownell) [inline roles/change-paths/tasks/main.yml]: So the example here would be: https://sites.stanford.edu/books/books/sites/default/files/book.jpg?

File: roles/css-injector/defaults/main.yml [hunk 15]
```diff
@@ -0,0 +1,9 @@
+---
+all_sar_commands: [ 'sar', 'sarl', 'sarm', 'sarv', 'sarjs' ]
```
Comment (kbrownell) [inline roles/css-injector/defaults/main.yml]: Not going to hold up the PR for this, but the format you're using in roles/change-paths/defaults/main.yml is easier to scan.  So thanks for thinking to make that change above.

File: roles/css-injector/tasks/main.yml [hunk 16]
```diff
@@ -0,0 +1,236 @@
...
+#   replace: url\1/
+#     Note: this will match single or double quotes:
+#       url('http://sites.stanford.edu/eao/
+#       url("http://sites.stanford.edu/eao/
+#   replace: url("/
+#     Note: this uses a backreference from the regexp match, and will replace with:
+#       url('/
+#       url("/
+- name: Remove absolute paths for url attributes in CSS Injector files
+  replace:
```
Comment (kbrownell) [inline roles/css-injector/tasks/main.yml]: So beautiful.

File: roles/css-injector/tasks/main.yml [hunk 17]
```diff
@@ -0,0 +1,236 @@
+---
+# Ansible role for changing paths in CSS Injector files.
+# https://github.com/SU-SWS/ansible-playbooks
+# ==================================================================
+#
+# This role runs a number of tasks to transform the database in such a way that
```
Comment (kbrownell) [inline roles/css-injector/tasks/main.yml]: You mention transformations are in the database, but these appear to all be file changes.  Am I reading the replace tasks wrong?

File: roles/define-paths/tasks/main.yml [hunk 18]
```diff
@@ -0,0 +1,97 @@
+---
+# Ansible role for setting a number of variables (mostly paths and URLs).
+# https://github.com/SU-SWS/ansible-playbooks
+# ==================================================================
+#
+# This role is required for "change-paths" and "css-injector".
```
Comment (kbrownell) [inline roles/define-paths/tasks/main.yml]: And for a whole lot of other roles, so I might be inclined to remove this line.

**Comment by jbickar** [comment]
Sweet

File: roles/css-injector/tasks/main.yml (hunk 17 above)
Comment (jbickar) [reply roles/css-injector/tasks/main.yml]: copypasta from `change-paths`

File: roles/change-paths/tasks/main.yml (hunk 14 above)
Comment (jbickar) [reply roles/change-paths/tasks/main.yml]: Yes, if that's the site. But the actual code replacement would be:
Before: `src="/books/sites/default/files/book.jpg`
After: `src="/sites/g/files/sbiybj5000986/f/book.jpg`

**Pull request merged**